src/AppInstallerRepositoryCore/PackageDependenciesValidation.cpp

```diff
@@ -0,0 +1,253 @@
+                const auto dependencies = GetDependencies(manifest, DependencyType::Package);
+                if (!dependencies.size())
+                {
+                    return depList;
+                }
+
+                std::for_each(dependencies.begin(), dependencies.end(), [&](Dependency dep) { depList.Add(dep);  });
+                return depList;
+            }
+
+            auto packageLatest = GetPackageLatestVersion(index, node.Id);
+            if (!packageLatest.has_value())
+            {
+                std::string error = ManifestError::MissingManifestDependenciesNode;
+                error.append(" ").append(node.Id);
+                dependenciesError.emplace_back(ValidationError(error));
+                foundErrors = true;
+                return depList;
+            }
+
+            if (node.MinVersion > packageLatest.value().second)
+            {
+                std::string error = ManifestError::NoSuitableMinVersion;
+                error.append(" ").append(node.Id);
+                dependenciesError.emplace_back(ValidationError(error));
+                foundErrors = true;
+                return depList;
+            }
+
+            auto packageLatestDependencies = index->GetDependenciesByManifestRowId(packageLatest.value().first);
+            std::for_each(
+                packageLatestDependencies.begin(),
+                packageLatestDependencies.end(),
+                [&](std::pair<std::pair<SQLite::rowid_t, Utility::NormalizedString>, SQLite::rowid_t> row)
+                {
+                    auto manifestRowId = index->GetManifestIdByKey(row.first.first, "", "");
+                    auto packageId = index->GetPropertyByManifestId(manifestRowId.value(), PackageVersionProperty::Id);
+                    Dependency dep(DependencyType::Package, packageId.value(), row.first.second);
+                    depList.Add(dep);
+                });
+
+            return depList;
+            });
+
+        graph.BuildGraph();
+
+        if (foundErrors)
+        {
+            THROW_EXCEPTION(ManifestException(std::move(dependenciesError), APPINSTALLER_CLI_ERROR_DEPENDENCIES_VALIDATION_FAILED));
+        }
+
+        if (graph.HasLoop())
+        {
+            std::string error = ManifestError::FoundLoop;
+            dependenciesError.emplace_back(error);
+            THROW_EXCEPTION(ManifestException(std::move(dependenciesError), APPINSTALLER_CLI_ERROR_DEPENDENCIES_VALIDATION_FAILED));
+        }
+
+        return true;
+    }
+
+    bool PackageDependenciesValidation::VerifyDependenciesStructureForManifestDelete(SQLiteIndex* index, const Manifest::Manifest manifest)
+    {
+        auto dependentsSet = index->GetDependentsById(manifest.Id);
+
+        if (!dependentsSet.size())
+        {
+            // all good this manifest is not a dependency of any manifest.
+            return true;
+        }
+
+        std::vector<std::pair<Manifest::Manifest, Utility::Version>> manifestToVersionPair;
```

Comment:
>Manifest::Manifest [](http://example.com/codeflow?start=30&length=18)

Why are you using a huge Manifest struct when all you care about is Id and Version?  And the Version in the manifest is the same as the version in the pair.

Reply:
No, they are not, the manifest struct presented an alternative to tuples or pairs, also I can't stress how much I don't like tuples because of their ugly, nonmaintainable and unreadable nature. It was a decision between creating a new struct for the sake of data transfer or using the manifest object.  Went with the latter, now switched to an internal struct for data transfer.